### PR TITLE
Persist remember me token

### PR DESCRIPTION
**Pull request by hedgehog1833**
Look at this beatiful PR :-)
The remember me token is now stored in a database table. The bug with the redirections is fixed. For a better user experience I added a redirect from the index page for already logged in users.
If you desperately want something to complain about: the two tests in the IndexControllerIT only test two of three endpoints. For some reason the empty_index endpoint gives a redirect in this test configuration. But after one day of building this I don't care anymore, espeacially as it works correctly in the Browser/Postman.

### CODE REVIEW

**Comment by hedgehog1833** [comment]
[Store remember me token in database](https://trello.com/c/XGhQPzn7/63-store-remember-me-token-in-database)

File: src/main/java/com/metalr2/security/SecurityConfig.java [hunk 1]
```diff
@@ -49,7 +55,7 @@ protected void configure(HttpSecurity http) throws Exception {
       .authorizeRequests()
         .antMatchers(Endpoints.AntPattern.ADMIN).hasRole(UserRole.ROLE_ADMINISTRATOR.getName())
         .antMatchers(Endpoints.AntPattern.RESOURCES).permitAll()
-        .antMatchers(Endpoints.AntPattern.AUTH_PAGES).anonymous()
+        .antMatchers(Endpoints.AntPattern.AUTH_PAGES).permitAll()
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/security/SecurityConfig.java]: Can a user who has been logged in now access the login page again or the register page?

File: src/main/java/com/metalr2/security/SecurityConfig.java [hunk 2]
```diff
@@ -59,8 +65,9 @@ protected void configure(HttpSecurity http) throws Exception {
         .failureHandler(new CustomAuthenticationFailureHandler())
       .and()
       .rememberMe()
-        .key(REMEMBER_ME_SECRET)
         .tokenValiditySeconds((int) Duration.ofDays(14).toSeconds())
+        .rememberMeServices(persistentTokenBasedRememberMeServices())
+        .key(REMEMBER_ME_SECRET)
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/security/SecurityConfig.java]: you can do it as follows:

```java
.rememberMe()
  .key(REMEMBER_ME_SECRET)
  .tokenRepository(persistentTokenRepository)
  .userDetailsService(userDetailsService)
```

After that you don't need the bean `persistentTokenBasedRememberMeServices`

File: src/main/java/com/metalr2/security/SecurityConfig.java [hunk 3]
```diff
@@ -78,4 +85,17 @@ protected void configure(AuthenticationManagerBuilder auth) throws Exception {
...
+  }
+
+  @Bean
+  public JdbcTokenRepositoryImpl jdbcTokenRepository() {
+    JdbcTokenRepositoryImpl jdbcTokenRepository = new JdbcTokenRepositoryImpl();
+    jdbcTokenRepository.setCreateTableOnStartup(false);
+    jdbcTokenRepository.setDataSource(dataSource);
+    return jdbcTokenRepository;
+  }
+
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/security/SecurityConfig.java]: You only need the following:

```java
@Bean
public PersistentTokenRespository persistentTokenRespository() {
  JdbcTokenRepositoryImpl jdbcTokenRepository = new JdbcTokenRepositoryImpl();
  jdbcTokenRepository.setCreateTableOnStartup(false);  // is this Default?
  jdbcTokenRepository.setDataSource(dataSource);

  return jdbcTokenRepository;
}
```

File: src/main/java/com/metalr2/security/SecurityConfig.java [hunk 4]
```diff
@@ -78,4 +85,17 @@ protected void configure(AuthenticationManagerBuilder auth) throws Exception {
     auth.userDetailsService(userService).passwordEncoder(bCryptPasswordEncoder);
   }
 
+  @Bean
+  public PersistentTokenBasedRememberMeServices persistentTokenBasedRememberMeServices() {
+    return new PersistentTokenBasedRememberMeServices(REMEMBER_ME_SECRET, userService, jdbcTokenRepository());
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/security/SecurityConfig.java]: You don't need separate bean `persistentTokenBasedRememberMeServices `

File: src/main/java/com/metalr2/web/controller/mvc/IndexController.java [hunk 5]
```diff
@@ -10,8 +11,10 @@
 public class IndexController {
 
   @GetMapping({Endpoints.Guest.INDEX, Endpoints.Guest.EMPTY_INDEX, Endpoints.Guest.SLASH_INDEX})
-  public ModelAndView showIndex() {
+  public ModelAndView showIndex(Authentication authentication) {
+    if (authentication != null && authentication.isAuthenticated()) {
+      return new ModelAndView("redirect:" + Endpoints.Frontend.HOME);
+    }
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/web/controller/mvc/IndexController.java]: I thought that you can define a corresponding setting in the `SecurityConfig` that, for example, always redirects to page XY by default after login.

Also: Suppose I call page `/my-artists`, but I am not logged in. I would expect to be redirected to the login page, log in and then be redirected back to the `/my-artists` page. But this does not happen. But I am of the opinion that this has worked last year. 

These should all be settings that you can make in the `SecurityConfig `.

File: src/main/java/com/metalr2/config/jpa/DatabaseConfiguration.java [hunk 6]
```diff
@@ -32,4 +44,13 @@
...
 
+  @Bean
+  public DataSource dataSource() {
+    DriverManagerDataSource dataSource = new DriverManagerDataSource();
+    dataSource.setDriverClassName(Objects.requireNonNull(environment.getProperty("spring.datasource.driver-class-name")));
+    dataSource.setUrl(environment.getProperty("spring.datasource.url"));
+    dataSource.setUsername(environment.getProperty("spring.datasource.username"));
+    dataSource.setPassword(environment.getProperty("spring.datasource.password"));
+    return dataSource;
+  }
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/config/jpa/DatabaseConfiguration.java]: Have you tried to autowire the `dataSource` Bean without explicit declaration?

Comment (hedgehog1833) [reply src/main/java/com/metalr2/config/jpa/DatabaseConfiguration.java]: Lol, that just f***ing works xD

**Comment by DanielW1987** [comment]
@hedgehog1833 
Please see my last commit. I think it's done now.

**Comment by hedgehog1833** [comment]
I implemented the interceptor and added some tests. I removed the check for the endpoint from the preHandle method as the interceptor is configured to run only on certain endpoints. This is tested in an integration test. preHandle functionality is tested in a unit test.

**Comment by DanielW1987** [comment]
In the tests, I have introduced the two classes `BaseWebMvcTest` and `BaseWebMvcTestWithSecurity`. This saves the whole boilerplate code around mocking. Please have a final look before merging.

**Pull request merged**